src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,316 @@
+        the version will be changed in place, no new version will be created
+    **kwargs : Any
+        Additional keyword arguments.
+
+    Returns
+    -------
+    str
+        The ID of the custom application source version.
+    """
+    return _get_or_create_custom_application_source_version(
+        endpoint=endpoint,
+        token=token,
+        custom_application_source_id=custom_application_source_id,
+        runtime_parameter_values=runtime_parameter_values,
+        previous_custom_application_source_version_id=previous_custom_application_source_version_id,
+        **kwargs,
+    )
+
+
+# def get_or_create_custom_application_source_version_from_previous(
+#     endpoint: str,
+#     token: str,
+#     custom_application_source_id: str,
+#     previous_custom_application_source_version_id: str,
+#     runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+#     **kwargs: Any,
+# ) -> str:
+#     """
+#     Get or create a custom application source version from a previous version.
+
+#     Parameters
+#     ----------
+#     custom_application_source_id : str
+#         The ID of the custom application source.
+#     previous_custom_application_source_version_id : str
+#         The ID of the previous custom application source version.
+#     runtime_parameter_values : Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]], optional
+#         A list of runtime parameter values. Each value can be either a `RuntimeParameterValue` object or a dictionary
+#         with parameter names as keys and parameter values as values. (default: None)
+#     **kwargs : Any
+#         Additional keyword arguments.
+
+#     Returns
+#     -------
+#     str
+#         The ID of the custom application source version.
+#     """
+#     return _get_or_create_custom_application_source_version(
+#         from_previous=True,
+#         endpoint=endpoint,
+#         token=token,
+#         custom_application_source_id=custom_application_source_id,
+#         previous_custom_application_source_version_id=previous_custom_application_source_version_id,
+#         runtime_parameter_values=runtime_parameter_values,
+#         **kwargs,
+#     )
+
+
+def get_or_create_qanda_app(
+    endpoint: Optional[str],
+    token: Optional[str],
+    deployment_id: str,
+    environment_id: str,
+    name: str,
+) -> str:
+    """Get or create a Q&A app.
+
+    Parameters
+    ----------
+    endpoint : Optional[str]
+        The DataRobot endpoint.
+    token : Optional[str]
+        The DataRobot API token.
+    deployment_id : str
+        The ID of the deployment.
+    environment_id : str
+        The ID of the environment.
+    name : str
+        The name of the Q&A app.
+
+    Returns
+    -------
+    str
+        The ID of the Q&A app.
+    """
+    app_token = get_hash(name, deployment_id, environment_id)
+
+    if endpoint is not None and token is not None:
```

Comment:
if aligned on above comment re: matching signature this goes away

Reply:
We need the from_previous - otherwise the environment is wiped